### PR TITLE
46 adaption of userservice to spec in miro board

### PR DESCRIPTION
**Pull request by WenzelSSE**

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,4 +1,4 @@
-APP_PORT=3000
+APP_PORT="your_Port"
```
Comment (Elscha) [inline .env:1]: Please revert this back to a default value: 3000

File: src/user/dto/learningProfile.dto.ts [hunk 2]
```diff
@@ -1,35 +1,51 @@
-import { IsDefined, IsNotEmpty } from 'class-validator';
+import { IsDefined, IsNotEmpty } from "class-validator";
 
-import { LearningProfile, Skill, LearningHistory } from '@prisma/client';
-import { OmitType } from '@nestjs/swagger';
-import { CareerProfileCreationDto } from './careerProfile-creation.dto';
-import { LearningProfileCreationDto } from './learningProfile-creation.dto';
+import { LearningProfile, Skill, LearningHistory } from "@prisma/client";
+import { OmitType } from "@nestjs/swagger";
```
Comment (Elscha) [inline src/user/dto/learningProfile.dto.ts:4]: Is this really used? I don't see this.

Please cleanup the code before merging. Delete all gryed imports.

**Review by Elscha — CHANGES_REQUESTED** [review]

**Comment by WenzelSSE** [comment]
^^please fix merge conflicts

File: src/user/dto/learningProfile.dto.ts (hunk 2 above)
Comment (Elscha) [reply src/user/dto/learningProfile.dto.ts:4]: I deleted OmitType, since it wasn't used in the class.

File: src/user/dto/qualification.dto.ts [hunk 3]
```diff
@@ -1,30 +1,29 @@
...
 export class QualificationDto extends QualificationCreationDto {
-  @IsNotEmpty()
-  id: string;
+    @IsNotEmpty()
+    id: string;
 
-  @IsDefined()
-  userCareerProfilId: string;
+    @IsDefined()
+    userCareerProfileId: string;
```
Comment (Elscha) [inline src/user/dto/qualification.dto.ts:12]: `userCareerProfileId `is labeld with `@IsDefined()`, but not set via the constructor.

File: src/user/dto/job-creation.dto.ts [hunk 4]
```diff
@@ -1,30 +1,52 @@
...
+    @IsOptional()
+    company?: Company;
+
+    @IsNotEmpty()
+    companyId: string;
+
+    @IsOptional()
+    userJobs?: CareerProfile[];
+
+    user: UserProfile;
```
Comment (Elscha) [inline src/user/dto/job-creation.dto.ts:29]: `JobCreationDto` is a request object, i.e., sent from outside and expects a `user` (internal `Prisma` object), which cannot send from the outside. Please resolve.

File: src/user/dto/personalizedLearningPath.dto.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            this.createdAt = createdAt;
+            this.updatedAt = updatedAt;
+        
+    }
+
+    static createFromDao(plp: PersonalizedLearningPath): PersonalizedLearningPathCreationDto {
+        const result: PersonalizedLearningPathDto = { 
+            id : plp.id,
+            pathTeachingGoals: plp.pathTeachingGoals,
+            unitSequence: plp.unitSequence,
```
Comment (Elscha) [inline src/user/dto/personalizedLearningPath.dto.ts:49]: `pathTeachingGoals` & `unitSequence` do not exist on `PersonalizedLearningPath`. Probably these are referenced fields of other tables (which may be selected via include `statements`. In this case, an intersection type is needed in Line 45, which specifies that these optiona files have been selected.

This is an example how to use an intersection type to specify that `nestedSkills` must be selected and `parentSkills` may be selected when passing a `Skill` object from `Prisma` to create a DTO:
```TypeScript
static createFromDao(
        skill: Skill & {
            nestedSkills: { id: string }[];
            parentSkills?: { id: string }[];
        },
    ): SkillDto {
        const dto: SkillDto = {
            id: skill.id,
            nestedSkills: skill.nestedSkills.map((element) => element.id),
            parentSkills: skill.parentSkills?.map((element) => element.id) || [],
            ...
        };
        return dto;
    }
```

File: src/user/dto/learningHistory.dto.ts [hunk 6]
```diff
@@ -1,31 +1,48 @@
...
+        return new LearningHistoryDto(
+            lh.id,
+            lh.startedLearningUnits,
+            lh.learnedSkills,
+            lh.user,
+            lh.userId,
+            lh.learningProfile,
+            //  lh.personalPaths,
+        );
+    }
```
Comment (Elscha) [inline src/user/dto/learningHistory.dto.ts:47]: `startedLearningUnits`, `learnedSkills`, `user` & `learningProfile` do not exist on `LearningHistory`. Probably these are referenced fields of other tables (which may be selected via `include` statements. In this case, an intersection type is needed in Line 37, which specifies that these optiona files have been selected.

This is an example how to use an intersection type to specify that `nestedSkills` must be selected and `parentSkills` may be selected when passing a `Skill` object from `Prisma` to create a DTO:
```TypeScript
static createFromDao(
        skill: Skill & {
            nestedSkills: { id: string }[];
            parentSkills?: { id: string }[];
        },
    ): SkillDto {
        const dto: SkillDto = {
            id: skill.id,
            nestedSkills: skill.nestedSkills.map((element) => element.id),
            parentSkills: skill.parentSkills?.map((element) => element.id) || [],
            ...
        };
        return dto;
    }
```

**Review by Elscha — CHANGES_REQUESTED** [review]

**Pull request closed without merging**